packer/linux/conf/buildkite-agent/scripts/stop-agent-gracefully

```diff
@@ -4,13 +4,44 @@ set -eu -o pipefail
 # Import BUILDKITE_AGENTS_PER_INSTANCE
 eval "$(cat /var/lib/buildkite-agent/cfn-env)"
 
-echo "Stopping buildkite-agent gracefully"
-systemctl stop "buildkite-agent"
+LIFECYCLE_TRANSITION=${1:-}
 
-# Need to ensure it's the buildkite-agent user, so it doesn't match this lifecycled handler script
-while pgrep -u buildkite-agent buildkite-agent > /dev/null; do
-  echo "Waiting for all buildkite-agent processes to have stopped..."
-  sleep 5
-done
+echo "Stopping buildkite-agent gracefully due to the lifecycle transition: ${LIFECYCLE_TRANSITION}"
 
-echo "All buildkite-agent processes have stopped"
+if [ $LIFECYCLE_TRANSITION = "ec2:SPOT_INSTANCE_TERMINATION" ]; then
+  # Send SIGTERM to the main buildkite-agent process in a non-blocking manner
+  # to start a graceful shutdown
+  systemctl kill --kill-who=main --signal=SIGTERM buildkite-agent
+
+  SPOT_TERMINATION_TIME=$(curl http://169.254.169.254/latest/meta-data/spot/termination-time 2>/dev/null)
```

Comment:
I think we'd want curl's `--fail` option, so that if the request fails it's picked up by bash's `set -e` and the script terminates, instead of proceeding with a blank `SPOT_TERMINATION_TIME`. (I _think_ terminating on failure is the lesser evil?)

> -f, --fail
> (HTTP)  Fail  silently  (no  output at all) on server errors. This is mostly done to better enable scripts etc to better deal with failed attempts. In normal cases when an HTTP server fails to deliver a document, it returns an HTML document stating so (which often also describes why and more). This flag will prevent curl from outputting that and return error 22.

Reply:
Good call @pda. I just realized lifecycled passes the termination time as the third argument to the script, so I updated this to rely on that.